src/core/service.c

```diff
@@ -1439,6 +1439,48 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static Service *service_get_triggering_service(Service *s) {
+        assert(s);
+
+        /* Return the service which triggered service 's', this means dependency
+         * types which include the UNIT_ATOM_PROPAGATE_EXIT_STATUS atom.
+         *
+         * N.B. if there are multiple services which could trigger 's' via OnFailure=
+         * or OnSuccess= then we return NULL. This is since we don't know from which
+         * one to propagate the exit status.
+         */
+
+        Unit *candidate = NULL;
+        Unit *other;
+
+        UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_PROPAGATE_EXIT_STATUS) {
+                assert(other->type == UNIT_SERVICE);
+
+                if (unit_has_dependency(other, UNIT_ATOM_ON_SUCCESS, UNIT(s)) ||
+                    unit_has_dependency(other, UNIT_ATOM_ON_FAILURE, UNIT(s))) {
+                    if (candidate) {
+                        log_unit_debug(UNIT(s), "multiple trigger source candidates for exit status propagation");
+                        return NULL;
+                    } else
+                        candidate = other;
+                }
+        }
+
+        if (!candidate)
+                return NULL;
+
+        if (unit_has_dependency(candidate, UNIT_ATOM_ON_SUCCESS, UNIT(s)) &&
+            SERVICE(candidate)->result == SERVICE_SUCCESS &&
+            unit_active_state(candidate) != UNIT_FAILED)
```

Comment:
hmm, should we really do these state checks here? I mean, if we don't, what's the worse that happens? that the unit specified on OnFailure= and OnSuccess= is started for some other reason and the env vars we pass are not actually about a failure or succcess but encode some other state? that wouldn't be much of a problem I think, after all they carry enough info for the invoked service to figure that out on it is own?

I'd really just say: when OnFailure=/OnSuccess= is configured onto some unit, then that unit will get the new env vars set, no matter what, regardless if actually started due to the OnFailur=/OnSuccess= or any other trigger. I think this would simplify things a lot, and not actually be surprising.

Reply:
Agreed, sounds much more sensible.